jvcprojector/projector.py

```diff
@@ -101,50 +100,140 @@ async def disconnect(self) -> None:
+            """Send commands and update the dictionary."""
+            cmd_vals = [JvcCommand(cmd, True) for cmd in commands.values()]
+            res = await self._send(cmd_vals)
+            # discard the command values and zip the keys with the responses
+            for (key, _), value in zip(commands.items(), res):
+                self._dict[key] = value
+
+        # Always get power state
+        await send_and_update({const.KEY_POWER: const.CMD_POWER})
+
+        # If power is on, get additional states
+        if self._dict.get(const.KEY_POWER) == const.ON:
+            await send_and_update(
+                {
+                    const.KEY_INPUT: const.CMD_INPUT,
+                    const.KEY_SOURCE: const.CMD_SOURCE,
+                    const.KEY_PICTURE_MODE: const.CMD_PICTURE_MODE,
+                    const.KEY_LOW_LATENCY: const.CMD_PICTURE_MODE_LOW_LATENCY,
+                    const.KEY_LASER_POWER: const.CMD_PICTURE_MODE_LASER_POWER,
+                    const.KEY_ANAMORPHIC: const.CMD_INSTALLATION_ANAMORPHIC,
+                    const.KEY_INSTALLATION_MODE: const.CMD_INSTALLATION_MODE,
+                }
+            )
+
+            # Check if there's a signal before getting signal-dependent states
+            if self._dict.get(const.KEY_SOURCE) == const.SIGNAL:
+                await send_and_update(
+                    {
+                        const.KEY_HDR: const.CMD_FUNCTION_HDR,
+                        const.KEY_HDMI_INPUT_LEVEL: const.CMD_INPUT_SIGNAL_HDMI_INPUT_LEVEL,
+                        const.KEY_HDMI_COLOR_SPACE: const.CMD_INPUT_SIGNAL_HDMI_COLOR_SPACE,
+                        const.KEY_COLOR_PROFILE: const.CMD_PICTURE_MODE_COLOR_PROFILE,
+                        const.KEY_GRAPHICS_MODE: const.CMD_PICTURE_MODE_GRAPHICS_MODE,
+                        const.KEY_COLOR_SPACE: const.CMD_FUNCTION_COLOR_SPACE,
+                    }
+                )
+
+            # NX9 and NZ model specific commands
+            if (
+                "NZ" in self._dict[const.KEY_MODEL]
+                or "NX9" in self._dict[const.KEY_MODEL]
+            ):
+                await send_and_update(
+                    {
+                        const.KEY_ESHIFT: const.CMD_PICTURE_MODE_8K_ESHIFT,
+                        const.KEY_CLEAR_MOTION_DRIVE: const.CMD_PICTURE_MODE_CLEAR_MOTION_DRIVE,
+                        const.KEY_MOTION_ENHANCE: const.CMD_PICTURE_MODE_MOTION_ENHANCE,
+                        const.KEY_LASER_VALUE: const.CMD_PICTURE_MODE_LASER_VALUE,
+                        const.KEY_LASER_TIME: const.CMD_FUNCTION_LASER_TIME,
+                        const.KEY_LASER_DIMMING: const.CMD_LASER_DIMMING,
+                    }
+                )
+
+            # HDR-specific commands
+            if self._dict.get("hdr") != const.HDR_CONTENT_SDR:
+                await send_and_update(
+                    {
+                        const.KEY_HDR_PROCESSING: const.CMD_PICTURE_MODE_HDR_PROCESSING,
```

Comment:
CMD_PICTURE_MODE_HDR_PROCESSING (PMHP) is timing out on my RS2000/NX7 100% of the time when there is no HDR source. 

```
2024-09-23 11:07:58.584 DEBUG (MainThread) [jvcprojector.device] Connecting to 192.168.0.30
2024-09-23 11:07:58.596 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJ_OK'
2024-09-23 11:07:58.597 DEBUG (MainThread) [jvcprojector.device] Handshake sending 'PJREQ'
2024-09-23 11:07:58.598 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJACK'
2024-09-23 11:07:58.598 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PW (b'?\x89\x01PW\n')'
2024-09-23 11:07:58.601 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PW\n'
2024-09-23 11:07:58.601 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PW1\n')
2024-09-23 11:07:58.803 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IP (b'?\x89\x01IP\n')'
2024-09-23 11:07:58.805 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IP\n'
2024-09-23 11:07:58.906 DEBUG (MainThread) [jvcprojector.device] Received ref b'6' (b'@\x89\x01IP6\n')
2024-09-23 11:07:59.107 DEBUG (MainThread) [jvcprojector.device] Sending ref 'SC (b'?\x89\x01SC\n')'
2024-09-23 11:07:59.111 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01SC\n'
2024-09-23 11:07:59.213 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01SC0\n')
2024-09-23 11:07:59.414 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMPM (b'?\x89\x01PMPM\n')'
2024-09-23 11:07:59.417 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:07:59.521 DEBUG (MainThread) [jvcprojector.device] Received ref b'03' (b'@\x89\x01PM03\n')
2024-09-23 11:07:59.723 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLL (b'?\x89\x01PMLL\n')'
2024-09-23 11:07:59.726 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:07:59.829 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PM1\n')
2024-09-23 11:08:00.030 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLP (b'?\x89\x01PMLP\n')'
2024-09-23 11:08:00.033 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:08:00.133 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
2024-09-23 11:08:00.334 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INVS (b'?\x89\x01INVS\n')'
2024-09-23 11:08:00.338 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:08:00.437 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01IN0\n')
2024-09-23 11:08:00.638 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INML (b'?\x89\x01INML\n')'
2024-09-23 11:08:00.642 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:08:00.742 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01IN1\n')
2024-09-23 11:08:00.944 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMHP (b'?\x89\x01PMHP\n')'
>>>>>>>> 2024-09-23 11:08:15.945 WARNING (MainThread) [jvcprojector.device] Response timeout for 'PMHP'
2024-09-23 11:08:16.147 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMCT (b'?\x89\x01PMCT\n')'
2024-09-23 11:08:16.153 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:08:16.253 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
2024-09-23 11:08:16.454 DEBUG (MainThread) [homeassistant.components.jvc_projector.coordinator] Finished fetching JVC Projector data in 17.870 seconds (success: True)
2024-09-23 11:08:18.455 DEBUG (MainThread) [jvcprojector.device] Disconnected
```

Then when an HDR signal becomes available it doesnt timeout, but gives the error `Index 3 out of range for command PMHP`. Annoyingly `3` is not documented in my V3.0 command reference/

```
2024-09-23 11:18:49.603 DEBUG (MainThread) [jvcprojector.device] Connecting to 192.168.0.30
2024-09-23 11:18:49.617 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJ_OK'
2024-09-23 11:18:49.617 DEBUG (MainThread) [jvcprojector.device] Handshake sending 'PJREQ'
2024-09-23 11:18:49.619 DEBUG (MainThread) [jvcprojector.device] Handshake received b'PJACK'
2024-09-23 11:18:49.619 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PW (b'?\x89\x01PW\n')'
2024-09-23 11:18:49.621 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PW\n'
2024-09-23 11:18:49.621 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PW1\n')
2024-09-23 11:18:49.823 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IP (b'?\x89\x01IP\n')'
2024-09-23 11:18:49.825 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IP\n'
2024-09-23 11:18:49.925 DEBUG (MainThread) [jvcprojector.device] Received ref b'6' (b'@\x89\x01IP6\n')
2024-09-23 11:18:50.126 DEBUG (MainThread) [jvcprojector.device] Sending ref 'SC (b'?\x89\x01SC\n')'
2024-09-23 11:18:50.132 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01SC\n'
2024-09-23 11:18:50.232 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01SC1\n')
2024-09-23 11:18:50.433 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMPM (b'?\x89\x01PMPM\n')'
2024-09-23 11:18:50.439 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:50.539 DEBUG (MainThread) [jvcprojector.device] Received ref b'0B' (b'@\x89\x01PM0B\n')
2024-09-23 11:18:50.741 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLL (b'?\x89\x01PMLL\n')'
2024-09-23 11:18:50.747 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:50.847 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PM1\n')
2024-09-23 11:18:51.048 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMLP (b'?\x89\x01PMLP\n')'
2024-09-23 11:18:51.053 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:51.153 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
2024-09-23 11:18:51.354 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INVS (b'?\x89\x01INVS\n')'
2024-09-23 11:18:51.361 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:18:51.461 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01IN0\n')
2024-09-23 11:18:51.663 DEBUG (MainThread) [jvcprojector.device] Sending ref 'INML (b'?\x89\x01INML\n')'
2024-09-23 11:18:51.669 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IN\n'
2024-09-23 11:18:51.769 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01IN1\n')
2024-09-23 11:18:51.971 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IFHR (b'?\x89\x01IFHR\n')'
2024-09-23 11:18:51.978 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IF\n'
2024-09-23 11:18:52.078 DEBUG (MainThread) [jvcprojector.device] Received ref b'2' (b'@\x89\x01IF2\n')
2024-09-23 11:18:52.279 DEBUG (MainThread) [jvcprojector.device] Sending ref 'ISIL (b'?\x89\x01ISIL\n')'
2024-09-23 11:18:52.286 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IS\n'
2024-09-23 11:18:52.386 DEBUG (MainThread) [jvcprojector.device] Received ref b'3' (b'@\x89\x01IS3\n')
2024-09-23 11:18:52.587 DEBUG (MainThread) [jvcprojector.device] Sending ref 'ISHS (b'?\x89\x01ISHS\n')'
2024-09-23 11:18:52.593 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IS\n'
2024-09-23 11:18:52.693 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01IS0\n')
2024-09-23 11:18:52.894 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMPR (b'?\x89\x01PMPR\n')'
2024-09-23 11:18:52.900 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:53.000 DEBUG (MainThread) [jvcprojector.device] Received ref b'24' (b'@\x89\x01PM24\n')
2024-09-23 11:18:53.201 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMGM (b'?\x89\x01PMGM\n')'
2024-09-23 11:18:53.207 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:53.306 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01PM1\n')
2024-09-23 11:18:53.508 DEBUG (MainThread) [jvcprojector.device] Sending ref 'IFXV (b'?\x89\x01IFXV\n')'
2024-09-23 11:18:53.511 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01IF\n'
2024-09-23 11:18:53.614 DEBUG (MainThread) [jvcprojector.device] Received ref b'1' (b'@\x89\x01IF1\n')
2024-09-23 11:18:53.815 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMHP (b'?\x89\x01PMHP\n')'
2024-09-23 11:18:53.818 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:53.922 DEBUG (MainThread) [jvcprojector.device] Received ref b'3' (b'@\x89\x01PM3\n')
2024-09-23 11:18:54.123 DEBUG (MainThread) [jvcprojector.device] Sending ref 'PMCT (b'?\x89\x01PMCT\n')'
2024-09-23 11:18:54.126 DEBUG (MainThread) [jvcprojector.device] Received ack b'\x06\x89\x01PM\n'
2024-09-23 11:18:54.226 DEBUG (MainThread) [jvcprojector.device] Received ref b'0' (b'@\x89\x01PM0\n')
>>>>>> 2024-09-23 11:18:54.427 WARNING (MainThread) [jvcprojector.command] Index 3 out of range for command PMHP
2024-09-23 11:18:54.428 DEBUG (MainThread) [homeassistant.components.jvc_projector.coordinator] Finished fetching JVC Projector data in 4.824 seconds (success: True)
^C2024-09-23 11:18:55.093 DEBUG (MainThread) [jvcprojector.device] Disconnected
```

Reply:
Ah at least for `Index 3 out of range error`, after looking at your lib I see I am missing the 0th index in my value definition like you have. 

What do you use as a reference for these values? The latest reference I know of is http://pro.jvc.com/pro/attributes/PRESENT/Manual/External%20Command%20Spec%20for%20D-ILA%20projector_V3.0.pdf. But that one has no 0th element for `hdr10_plus` like you have.

Reply:
Ok after further digging, line 189 below is whats allowing the call to PMHP to erroniously happen, because in my case the HDR value is `none(F)`, not `SDR(0)`.

```python
if self._dict.get("hdr") != const.HDR_CONTENT_SDR:
```

Reply:
updated to check for none as well